File: backend/routes/budgets.py
"""
Budget management routes.
"""
from fastapi import APIRouter, HTTPException, Depends, Query, status
from sqlalchemy.orm import Session
from database import get_db
from auth import get_current_user
import crud
from schemas import BudgetCreate, BudgetUpdate, BudgetResponse

router = APIRouter(prefix="/budgets", tags=["budgets"])


@router.post("", response_model=BudgetResponse, status_code=status.HTTP_201_CREATED)
def create_budget(
    budget: BudgetCreate,
    user_id: int = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new budget."""
    return crud.create_budget(db, budget, user_id)


@router.get("", response_model=list[BudgetResponse])
def get_budgets(
    user_id: int = Depends(get_current_user),
    db: Session = Depends(get_db),
    year: int = Query(...),
    month: int = Query(..., ge=1, le=12),
):
    """Get budgets for a specific month/year."""
    return crud.get_user_budgets(db, user_id, year, month)


@router.get("/{budget_id}", response_model=BudgetResponse)
def get_budget(
    budget_id: int,
    user_id: int = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get a specific budget."""
    budget = crud.get_budget(db, budget_id, user_id)
    if not budget:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Budget not found"
        )
    return budget


@router.put("/{budget_id}", response_model=BudgetResponse)
def update_budget(
    budget_id: int,
    budget_update: BudgetUpdate,
    user_id: int = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update a budget."""
    budget = crud.update_budget(db, budget_id, user_id, budget_update)
    if not budget:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Budget not found"
        )
    return budget


@router.delete("/{budget_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_budget(
    budget_id: int,
    user_id: int = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a budget."""
    success = crud.delete_budget(db, budget_id, user_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Budget not found"
        )
    return None

File: frontend/src/api/budgets.ts
/**
 * API endpoints for budgets
 */
import { api } from '@/lib/apiClient';
import { Budget } from '@/types';

export const budgetsApi = {
  async createBudget(data: {
    category: string;
    monthly_limit: number;
    month: number;
    year: number;
  }): Promise<Budget> {
    const response = await api.post('/budgets', data);
    return response.data;
  },

  async getBudgets(year: number, month: number): Promise<Budget[]> {
    const response = await api.get(`/budgets?year=${year}&month=${month}`);
    return response.data;
  },

  async getBudget(id: number): Promise<Budget> {
    const response = await api.get(`/budgets/${id}`);
    return response.data;
  },

  async updateBudget(id: number, data: { monthly_limit: number }): Promise<Budget> {
    const response = await api.put(`/budgets/${id}`, data);
    return response.data;
  },

  async deleteBudget(id: number): Promise<void> {
    await api.delete(`/budgets/${id}`);
  },
};
